qa/tasks/cephfs/test_cephfs_shell.py

```diff
@@ -293,11 +318,125 @@ def test_cd_with_args(self):
         output = self.get_cephfs_shell_script_output(script)
         self.assertEqual(output, expected_cwd)
 
+class TestDU(TestCephFSShell):
+    CLIENTS_REQUIRED = 1
+
+    def test_du_works_for_regfiles(self):
+        regfilename = 'some_regfile'
+        regfile_abspath = path.join(self.mount_a.mountpoint, regfilename)
+        sudo_write_file(self.mount_a.client_remote, regfile_abspath, 'somedata')
+
+        size = humansize(self.mount_a.stat(regfile_abspath)['st_size'])
+        expected_output = r'{}{}{}'.format(size, " +", regfilename)
+
+        du_output = self.get_cephfs_shell_cmd_output('du ' + regfilename)
+        if sys_version_info.major > 3:
+            self.assertRegex(expected_output, du_output)
+        elif sys_version_info.major < 3:
+            assert re_search(expected_output, du_output) != None, "\n" + \
```

Comment:
Is the reason for the regex search to filter out colored output?

Reply:
Yep. `shell.color = Never` stopped colouring on my machine but that didn't happen on teuthology tests. Perhaps, that has something to do with `cmd2`'s version teuthology machine. I wanted to dig in further but it would take too much time and I did ask you about `cmd2`'s version via email but I haven't got a reply yet. :P

Also to avoid any coincidence, I've replaced the directory name with a longer one.

Reply:
Since, until last time I was seeing the issue, should I create a tracker ticket for that? I am unsure if it would helpful since neither I've got any idea of the cause behind and nor I can reproduce it locally.

Reply:
> I wanted to dig in further but it would take too much time and I did ask you how to find cmd2's version via email but I haven't got a reply yet.

`cmd2.__version__` seems to have the information.

Reply:
The `cmd2` available on teuthology will be older. Some of these meta commands probably do not exist.

There must be a way to disable coloring somehow in cmd2, even for those versions. Please look into it. Using regex to remove the coloring is not sustainable nor robust.

Reply:
> > I wanted to dig in further but it would take too much time and I did ask you how to find cmd2's version via email but I haven't got a reply yet.
> 
> `cmd2.__version__` seems to have the information.

I needed to know the version of cmd2 on teuthology machine. When I wrote that comment, it was close to the email I had sent you on this topic.  What I wanted to do back then was verify that `set colors Never` did turn the coloring off on the version teuthology machine uses. But later I figured another way to circumvent the error.

cmd2.Cmd.colors needs to be set to`Never` to turn the colouring off. Since, cephfs-shell can't pick `colors`'s values from ceph.conf, I don't think there's any other way left.

> There must be a way to disable coloring somehow in cmd2, even for those versions. Please look into it. Using regex to remove the coloring is not sustainable nor robust.

My plan is to keep the test as it is for now and later make cephfs-shell pick up values fron ceph.conf and then rewrite the test.

There's one more way... Would regex search on the output of `cephfs-shell -c ceph.conf 'set colors Never, ls'` be better than what we have now?

Reply:
Would suggestion above work?

Reply:
I believe the core problem is that running `cephfs-shell` on the remote this way results in `stdout` being a terminal. That means that colors are not suppressed by default. According to the documentation [1], the colors would be suppressed if the output is a pipe or file.

I recommend simply changing `run_cephfs_shell_cmd` to execute instead:

`cephfs-shell ... | cat`

That way cephfs-shell always sees a pipe for stdout and colors will therefore be suppressed.

[1] https://buildmedia.readthedocs.org/media/pdf/cmd2/latest/cmd2.pdf

Reply:
The stdout we want would be lost in that case.

Reply:
Adding the code to make cephfs-shell read cmd2's default variables from ceph.conf - https://github.com/ceph/ceph/pull/29964